yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryProvider.scala

```diff
@@ -0,0 +1,1015 @@
+
+  private val refreshCount = new AtomicLong(0)
+  private val refreshFailedCount = new AtomicLong(0)
+
+  /**
+   * Health marker
+   */
+  private val healthy = new AtomicBoolean(false)
+
+  /**
+   * Enabled flag
+   */
+  private val _enabled = timelineServiceEnabled(yarnConf)
+
+  /**
+   * Atomic boolean used to signal to the refresh thread that it
+   * must exit its loop.
+   */
+  private val stopRefresh = new AtomicBoolean(false)
+
+  /**
+   * refresher
+   */
+  val refresher = new Refresher()
+
+  /**
+   * Initialize the provider
+   */
+  init()
+
+  /**
+   * Check the configuration and log whether or not it is enabled;
+   * if it is enabled then the URL is logged too.
+   */
+  private def init(): Unit = {
+    if (!enabled) {
+      logError(YarnHistoryProvider.TEXT_SERVICE_DISABLED)
+    } else {
+      logInfo(YarnHistoryProvider.TEXT_SERVICE_ENABLED)
+      logInfo(YarnHistoryProvider.KEY_SERVICE_URL + ": " + timelineEndpoint)
+      logDebug(sparkConf.toDebugString)
+      // get the thread time
+      logInfo(s"refresh interval $refreshInterval milliseconds")
+      if (refreshInterval < 0) {
+        throw new Exception(YarnHistoryProvider.TEXT_INVALID_UPDATE_INTERVAL +
+            s": ${refreshInterval/1000}")
+      }
+      startRefreshThread()
+    }
+  }
+
+
+  /**
+   * Stop the service. After this point operations will fail.
+   */
+  override def stop(): Unit = {
+    logDebug(s"Stopping $this")
+    // attempt to stop the refresh thread
+    if (!stopRefreshThread()) {
+      closeQueryClient()
+    }
+
+  }
+
+  /**
+   * Close the query client
+   */
+  def closeQueryClient(): Unit = {
+    // and otherwise, stop the query client
+    logDebug("Stopping Timeline client")
+    timelineQueryClient.close()
+  }
+
+  /**
+   * Is the timeline service (and therefore this provider) enabled.
+   * (override point for tests).
+   *
+   * Important: this is called during construction, so test-time subclasses
+   * will be invoked before their own construction has taken place.
+   * Code appropriately.
+   * @return true if the provider/YARN configuration enables the timeline
+   *         service.
+   */
+  def enabled: Boolean = {
+    _enabled
+  }
+  
+  /**
+   * Get the timeline query client. Used internally to ease testing
+   * @return the client.
+   */
+  def getTimelineQueryClient(): TimelineQueryClient = {
+    timelineQueryClient
+  }
+
+  /**
+   * Set the last exception
+   * @param ex exception seen
+   */
+  private def setLastFailure(ex: Throwable): Unit = {
+    setLastFailure(ex, now())
+  }
+
+  /**
+   * Set the last exception
+   * @param ex exception seen
+   * @param timestamp the timestamp of the failure
+   */
+  private def setLastFailure(ex: Throwable, timestamp: Long): Unit = {
+    this.synchronized {
+      lastFailureCause = Some(ex, new Date(timestamp))
+    }
+  }
+
+  /**
+   * Reset the failure info
+   */
+  private def resetLastFailure(): Unit = {
```

Comment:
a shorthand version for this would be:

    private def resetLastFailure(): Unit = synchronized {
      // code
    }